src/controller.js

```diff
@@ -63,90 +39,69 @@ function buildData(dataset, mainRect, dpr) {
           subRect.y += font.lineHeight + padding * 2;
           subRect.h -= font.lineHeight + padding * 2;
         }
-        ret.push(...recur(gidx + 1, rasterizeRect(subRect, dpr), sq.g, sq.s));
+        ret.push(...recur(gidx + 1, subRect, sq.g, sq.s));
       });
     }
     return ret;
   }
 
-  if (!tree.length && dataset.data.length) {
-    tree = dataset.tree = dataset.data;
-  }
-
   return glen
     ? recur(0, mainRect)
     : squarify(tree, mainRect, key, dpr);
 }
 
-function getMinMax(data, useTree) {
-  const vMax = useTree ? 1 : maxValue(data);
-  const vMin = useTree ? 0 : minValue(data, vMax);
-  return {vMin, vMax};
-}
-
-function getArea({xScale, yScale}, data, rtl, useTree) {
-  const {vMin, vMax} = getMinMax(data, useTree);
-  const xMin = xScale.getPixelForValue(0);
-  const xMax = xScale.getPixelForValue(1);
-  const yMin = yScale.getPixelForValue(vMin);
-  const yMax = yScale.getPixelForValue(vMax);
-  return {x: xMin, y: yMax, w: xMax - xMin, h: yMin - yMax, rtl};
-}
-
 export default class TreemapController extends DatasetController {
   constructor(chart, datasetIndex) {
     super(chart, datasetIndex);
 
     this._rect = undefined;
     this._key = undefined;
     this._groups = undefined;
-    this._useTree = undefined;
   }
 
   initialize() {
     this.enableOptionSharing = true;
     super.initialize();
   }
 
-  /**
-   * @todo: Remove with https://github.com/kurkle/chartjs-chart-treemap/issues/137
-   */
-  updateRangeFromParsed(range, scale) {
-    if (range.updated) {
-      return;
-    }
-    range.updated = true;
-    if (scale.axis === 'x') {
-      range.min = 0;
-      range.max = 1;
-      return;
-    }
-    const dataset = this.getDataset();
-    const {vMin, vMax} = getMinMax(dataset.data, this._useTree);
-    range.min = vMin;
-    range.max = vMax;
+  // eslint-disable-next-line no-unused-vars
+  getMinMax(scale, canStack) {
+    return {
+      min: 0,
+      max: scale.axis === 'x' ? this.chart.width : this.chart.height
+    };
+  }
+
+  configure() {
+    super.configure();
+    // console.log(this.getMeta());
   }
 
   update(mode) {
     const meta = this.getMeta();
     const dataset = this.getDataset();
-    const dpr = this.chart.currentDevicePixelRatio;
+    const area = this.chart.chartArea;
 
-    if (!defined(this._useTree)) {
-      this._useTree = !!dataset.tree;
-    }
     const groups = dataset.groups || (dataset.groups = []);
     const key = dataset.key || '';
     const rtl = !!dataset.rtl;
+    const tree = dataset.tree || dataset.data;
+
+    const mainRect = {
```

Comment:
I think we could use the iScale and vScale dimensions instead of chartArea. This will enable the stacked axes and multi treemap datasets in the same chart. I don't see any performance issue.

Reply:
Updated to use x/yScale